apis/python/src/tiledbsoma/_dataframe.py

```diff
@@ -401,6 +401,37 @@ def get_enumeration_values(
 
         return self._handle.get_enumeration_values(column_names)
 
+    def extend_enumeration_values(
+        self: DataFrame, values: dict[str, pa.Array], deduplicate: bool = False
+    ) -> None:
+        """Extend enumeration values for each column defined in `values`.
+        Raises ``ValueError`` if any of the the specified column names is not in
+        the schema, or if any is not of Arrow dictionary type. All the values
+        must provided must not already exist in the schema (see the output of
+        ``get_enumeration_values``), unless ``deduplicate=True`` is provided."""
+        self.verify_open_for_writing()
+
+        # These assertions could be done in C++. However, it's easier here
+        # to do the exception-type multiplexing, raising ValueError for one
+        # thing, TileDBSOMAError for another.
+        for column_name in values.keys():
+            # As with get_enumeration_values: we are trusting pyarrow to raise
+            # KeyError, and raise it with a sufficiently clear error message,
+            # when the column name is not present within the schema.
+            field = self.schema.field(column_name)
+            if not pa.types.is_dictionary(field.type):
+                raise KeyError(
+                    f"schema column name '{column_name}' is not of dictionary type"
+                )
+            if pa.types.is_dictionary(values[column_name].type):
+                raise ValueError(
+                    f"value column name '{column_name}' is of dictionary type: pass its dictionary array instead"
+                )
+                # TO DISCUSS IN CODE REVIEW: alternatively:
```

Comment:
once we have discussed sufficiently in code review, would be good to remove this comment

Reply:
https://github.com/single-cell-data/TileDB-SOMA/pull/3815/commits/fc578bd9e804b433f7087cd9928bdb01941cd091